src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -70,43 +70,50 @@ UInt64 GinIndexPostingsBuilder::serialize(WriteBuffer & buffer)
     }
 
     const bool compress = cardinality >= ROARING_ENCODING_COMPRESSION_CARDINALITY_THRESHOLD;
-    const UInt64 uncompressed_size = rowid_bitmap.getSizeInBytes();
+    const UInt64 uncompressed_size = rowids.getSizeInBytes();
 
-    auto buf = std::make_unique<char[]>(uncompressed_size);
-    rowid_bitmap.write(buf.get());
+    std::vector<char> buf(uncompressed_size);
+    rowids.write(buf.data());
 
-    UInt64 header = (uncompressed_size << 1);
-    header <<= 1;
+    UInt64 header = uncompressed_size;
     if (compress)
     {
         Memory<> memory;
         const auto & codec = GinIndexCompressionFactory::zstdCodec();
         memory.resize(codec->getCompressedReserveSize(static_cast<UInt32>(uncompressed_size)));
-        auto compressed_size = codec->compress(buf.get(), static_cast<UInt32>(uncompressed_size), memory.data());
+        auto compressed_size = codec->compress(buf.data(), static_cast<UInt32>(uncompressed_size), memory.data());
 
-        header |= 0x2; /// zstd compressed + use roaring
+        header = (header << 2) | (ROARING_COMPRESSED_MASK << 1) | ROARING_CONTAINER_MASK;
 
         writeVarUInt(header, buffer);
         writeVarUInt(compressed_size, buffer);
         buffer.write(memory.data(), compressed_size);
 
         return getLengthOfVarUInt(header) + getLengthOfVarUInt(compressed_size) + compressed_size;
     }
+    else
+    {
+        header = (header << 2) | (ROARING_UNCOMPRESSED_MASK << 1) | ROARING_CONTAINER_MASK;
 
-    header |= 0x0; /// uncompressed + use roaring
-
-    writeVarUInt(header, buffer);
-    buffer.write(buf.get(), uncompressed_size);
+        writeVarUInt(header, buffer);
+        buffer.write(buf.data(), uncompressed_size);
 
-    return getLengthOfVarUInt(header) + uncompressed_size;
+        return getLengthOfVarUInt(header) + uncompressed_size;
+    }
 }
 
 GinIndexPostingsListPtr GinIndexPostingsBuilder::deserialize(ReadBuffer & buffer)
 {
+    /**
+     * Header value maps into following states:
+     * The lowest bit indicates if values are stored as an array or Roaring bitmap
+     * In case of array container, the rest of the bits is the number of entries in the array.
+     * In case of Roaring bitmap, the second lowest bit indicates if Roaring bitmap is compressed or uncompressed, the rest of the bits is the uncompressed size.
+     */
```

Comment:
Isn't is simpler to use the highest bits for flags instead of the lower ones?
If I understand correctly we should have enough empty bits in the highest part to use as future flags if we need them without breaking backward compatibility.

So, we could extract both parts safety with less manipulation and simpler logic. WDYT?